### PR TITLE
Allow Run to return an error to be emitted.

### PR DESCRIPTION
**Pull request by clkunzang**
This addresses #86 straightforwardly.

### CODE REVIEW

**Comment by Stebalien** [comment]
There's also a call to Run in `command.go`.

File: command_test.go [hunk 1]
```diff
@@ -401,3 +403,35 @@ func TestCancel(t *testing.T) {
...
 }
+
+type testEmitterWithError testing.T
+
+func (s *testEmitterWithError) Close() error {
+	return nil
+}
+
+func (s *testEmitterWithError) SetLength(_ uint64)                              {}
+func (s *testEmitterWithError) SetError(err interface{}, code cmdkit.ErrorType) {}
```
Comment (Stebalien) [inline command_test.go]: We should record the error here and make sure it's set before the test ends.

Comment (clkunzang) [reply command_test.go]: Done.

File: command_test.go [hunk 2]
```diff
@@ -2,10 +2,15 @@ package cmds
 
 import (
 	"context"
+	"errors"
 	"io"
 	"testing"
 	"time"
 
+	"sync"
```
Comment (Stebalien) [inline command_test.go]: We generally group like imports together. Standard, then local (this repo), then remote (other repos) with newlines between).

File: command_test.go [hunk 3]
```diff
@@ -401,3 +407,57 @@ func TestCancel(t *testing.T) {
...
+
+func (s *testEmitterWithError) Close() error {
+	return nil
+}
+
+func (s *testEmitterWithError) SetLength(_ uint64) {}
+
+func (s *testEmitterWithError) SetError(err interface{}, code cmdkit.ErrorType) {
+	errorCount++
+	fmt.Printf("errorCount: %d\n", errorCount)
```
Comment (Stebalien) [inline command_test.go]: Tests generally shouldn't use printf.

File: command_test.go [hunk 4]
```diff
@@ -401,3 +407,57 @@ func TestCancel(t *testing.T) {
...
+type testEmitterWithError testing.T
+
+func (s *testEmitterWithError) Close() error {
+	return nil
+}
+
+func (s *testEmitterWithError) SetLength(_ uint64) {}
+
+func (s *testEmitterWithError) SetError(err interface{}, code cmdkit.ErrorType) {
+	errorCount++
```
Comment (Stebalien) [inline command_test.go]: We *definitely* shouldn't be using a global here. Instead, I'd stash the error in the test emitter (or set a flag or something).

File: command_test.go (hunk 3 above)
Comment (clkunzang) [reply command_test.go]: Grrr... Debugging, sorry.

**Comment by Stebalien** [comment]
LGMT. Thanks!

File: examples/adder/cmd.go [hunk 5]
```diff
@@ -29,36 +29,37 @@ var RootCmd = &cmds.Command{
...
-			Run: func(req *cmds.Request, re cmds.ResponseEmitter, env cmds.Environment) {
+			Run: func(req *cmds.Request, re cmds.ResponseEmitter, env cmds.Environment) error {
 				sum := 0
 
 				for i, str := range req.Arguments {
 					num, err := strconv.Atoi(str)
 					if err != nil {
 						re.SetError(err, cmdkit.ErrNormal)
-						return
+						return err
```
Comment (dignifiedquire) [inline examples/adder/cmd.go]: shouldn't this remove the above `SetError` call, as otherwise these get emitted twice?

Comment (clkunzang) [reply examples/adder/cmd.go]: Good catch. I'll fix those.

**Comment by whyrusleeping** [comment]
Hey guys, it would be great to not push breaking changes haphazardly. I'm debating reverting this...

**Comment by whyrusleeping** [comment]
Yeah, i'm reverting this. Please try again when there is a reasonably clear upgrade path, and this has been tested in go-ipfs.

**Comment by whyrusleeping** [comment]
master is now 6d7f4fc86b88d912f4f836847b819a50d75c11c6

**Comment by Stebalien** [comment]
@whyrusleeping I've already volunteered to fix go-ipfs and this *does* make returning errors *much* easier.

For old commands, the upgrade path is to modify the shim. For new commands, I plan on fixing this with sed and some manual labor (shouldn't be that hard).

I agree this should have been discussed in an issue and I dropped the ball there. However, I *do* remember discussing this with you in the past and didn't here any objections other than "juan had reasons".

**Comment by whyrusleeping** [comment]
@Stebalien I'm okay with this change in general, but making a breaking change to a library like this should be accompanied by any fixes required downstream. Look at it this way, if we find a critical bug that needs fixing immediately in this package before these changes are made, we have to either revert these changes and make the fix, or we have to do a whole bunch of extra changes (and test them) to get that fix out.

This is especially important given how unstable this particular codebase has been lately.

**Comment by Stebalien** [comment]
Ah. Yeah, fair enough.

**Pull request merged**